src/binder/query/query_graph_label_analyzer.cpp

```diff
@@ -94,49 +95,231 @@ void QueryGraphLabelAnalyzer::pruneNode(const QueryGraph& graph, NodeExpression&
+    std::vector<table_id_t> nodeTableIDs;
+    for (auto node_table_entry : catalog->getNodeTableEntries(tx)) {
+        nodeTableIDs.push_back(node_table_entry->getTableID());
+    }
+    return nodeTableIDs;
+}
+
+std::unordered_set<TableCatalogEntry*> QueryGraphLabelAnalyzer::mergeTableIDs(
+    const std::vector<table_id_set_t>& v1, const std::vector<table_id_set_t>& v2) const {
+    std::unordered_set<table_id_t> temp;
+    for (auto tableIDs : v1) {
+        temp.insert(tableIDs.begin(), tableIDs.end());
+    }
+    for (auto tableIDs : v2) {
+        temp.insert(tableIDs.begin(), tableIDs.end());
+    }
+    std::unordered_set<TableCatalogEntry*> ans;
+    for (table_id_t tableID : temp) {
+        ans.emplace(catalog->getTableCatalogEntry(tx, tableID));
+    }
+    return ans;
+}
+
+static std::vector<catalog::TableCatalogEntry*> intersectEntries(
+    std::vector<catalog::TableCatalogEntry*> v1, std::vector<catalog::TableCatalogEntry*> v2) {
+    std::sort(v1.begin(), v1.end());
+    std::sort(v2.begin(), v2.end());
+    std::vector<catalog::TableCatalogEntry*> intersection;
+    std::set_intersection(v1.begin(), v1.end(), v2.begin(), v2.end(),
+        std::back_inserter(intersection));
+    return intersection;
+}
+
+static bool isSameTableCatalogEntryVector(std::vector<TableCatalogEntry*> v1,
+    std::vector<TableCatalogEntry*> v2) {
+    auto compareFunc = [](TableCatalogEntry* a, TableCatalogEntry* b) {
+        return a->getTableID() < b->getTableID();
+    };
+    std::sort(v1.begin(), v1.end(), compareFunc);
+    std::sort(v2.begin(), v2.end(), compareFunc);
+    return std::equal(v1.begin(), v1.end(), v2.begin(), v2.end());
+}
+
+void QueryGraphLabelAnalyzer::pruneRel(RelExpression& rel) const {
+    auto srcTableIDSet = rel.getSrcNode()->getTableIDsSet();
+    auto dstTableIDSet = rel.getDstNode()->getTableIDsSet();
+    if (rel.isRecursive()) {
+        auto nodeTableIDs = getNodeTableIDs();
+        // there is no label on both sides
+        if (rel.getUpperBound() == 0 || (srcTableIDSet.size() == dstTableIDSet.size() &&
+                                            dstTableIDSet.size() == nodeTableIDs.size())) {
+            return;
+        }
+
+        auto [stepFromLeftTableIDs, stepFromRightTableIDs] =
+            pruneRecursiveRel(rel.getEntries(), srcTableIDSet, dstTableIDSet, rel.getLowerBound(),
+                rel.getUpperBound(), rel.getDirectionType());
+        auto recursiveInfo = rel.getRecursiveInfo();
+        recursiveInfo->stepFromLeftActivationRelInfos = stepFromLeftTableIDs;
+        recursiveInfo->stepFromRightActivationRelInfos = stepFromRightTableIDs;
+        // todo we need reset rel entries?
+        auto temp = mergeTableIDs(stepFromLeftTableIDs, stepFromRightTableIDs);
+        std::vector<TableCatalogEntry*> newRelEntries{temp.begin(), temp.end()};
+        if (!isSameTableCatalogEntryVector(newRelEntries, rel.getEntries())) {
```

Comment:
Shall we do l284-l312 as a separate PR because

- it can be viewed as another optimization after pruning recursive rel labels; and
- we will also need to get a benchmark number to make sure there are performance improvements for node pruning.